fiftyone/plugins/secrets.py

```diff
@@ -128,32 +146,36 @@ def __getitem__(self, key):
                 self.__secrets[key] = val
         return val
 
+    def __setattr__(self, key, value):
+        if re.search("_SecretsDictionary__(secrets|required_keys)$", key):
+            raise KeyError("Cannot mutate hidden properties")
+        else:
+            super().__setattr__(key, value)
+
     def __setitem__(self, key, value):
         raise RuntimeError("Setting values is not allowed")
 
     def __deepcopy__(self, memodict={}):
-        logging.warning(
-            "Copying the SecretDictionary values is not " "allowed."
-        )
+        logging.warning("Copying the SecretsDictionary values is not allowed.")
 
         return {k: None for k in self.__secrets.keys()}
 
     def __dict__(self):
         return {k: True for k, v in self.__secrets.items() if v is not None}
 
     def copy(self):
-        logging.warning(
-            "Copying the SecretDictionary values is not " "allowed."
-        )
+        logging.warning("Copying the SecretsDictionary values is not allowed.")
         return self.__deepcopy__()
 
     def keys(self):
         return [k for k, v in self.__secrets.items() if v is not None]
 
     def values(self):
-        # Override values() to prevent iterating through plaintext values
-        raise RuntimeError("Iteration through values is not allowed")
+        # Override values() to ensure that resolvable secrets are always returned upon iteration
```

Comment:
My suggestion to inherit from `Mapping` had two benefits:
- it automatically handles throwing errors when mutation methods like `__setitem__` and `update()` are called
- you only have to implement `__getitem__`, `__len__`, and `__iter__` as described [here](https://github.com/voxel51/fiftyone/pull/3820#pullrequestreview-1740197017) and then other methods like `keys()`, `values()`, and `items()` will automatically use those in the correct ways.

In other words, once you fix your `__iter__` implementation I believe `keys()`, `values()`, and `items()` can just be deleted.